sdk/core/azure-core-amqp/inc/azure/core/amqp/connection_string_credential.hpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-Licence-Identifier: MIT
+
+#pragma once
+
+#include "network/sasl_transport.hpp"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace Core { namespace Amqp { namespace _internal {
+  //
+  // A ServiceBus connection string has the following format:
+  // "Endpoint=sb://<namespace>.servicebus.windows.net/;SharedAccessKeyName=<KeyName>;SharedAccessKey=<KeyValue>;EntityPath=<entity>"
+  //
+
+  enum class CredentialType
+  {
+    None,
+    SaslPlain,
+    ServiceBusSas,
+    BearerToken,
+  };
+
+  class ConnectionStringCredential {
+  public:
+    ConnectionStringCredential(const std::string& connectionString, CredentialType credentialType)
+        : m_credentialType{credentialType}
+    {
+      ParseConnectionString(connectionString);
+    }
+    virtual ~ConnectionStringCredential() = default;
+
+    // Prevent tearing of the ConnectionStringCredential.
+    ConnectionStringCredential(ConnectionStringCredential const&) = delete;
+    ConnectionStringCredential& operator=(ConnectionStringCredential const&) = delete;
+
+    virtual std::shared_ptr<Network::_internal::Transport> GetTransport() const = 0;
+    CredentialType GetCredentialType() const { return m_credentialType; }
+    std::string const& GetEndpoint() const { return m_endpoint; }
+    std::string const& GetSharedAccessKeyName() const { return m_sharedAccessKeyName; }
+    std::string const& GetSharedAccessKey() const { return m_sharedAccessKey; }
+    std::string const& GetEntityPath() const { return m_entityPath; }
+    std::string const& GetHostName() const { return m_hostName; }
+    uint16_t GetPort() const { return m_port; }
+
+  private:
+    void ParseConnectionString(const std::string& connectionString);
+    const CredentialType m_credentialType;
+    std::string m_endpoint;
+    std::string m_sharedAccessKeyName;
+    std::string m_sharedAccessKey;
+    std::string m_uri;
+    std::string m_hostName;
+    uint16_t m_port;
+
+  protected:
+    std::string m_entityPath;
+  };
+
+  class ServiceBusSasConnectionStringCredential final : public ConnectionStringCredential {
+  public:
+    ServiceBusSasConnectionStringCredential(
+        const std::string& connectionString,
+        const std::string& entityPath = {})
+        : ConnectionStringCredential(connectionString, CredentialType::ServiceBusSas)
+    {
+      if (m_entityPath.empty())
+      {
+        m_entityPath = entityPath;
+      }
+      else if (!entityPath.empty() && m_entityPath != entityPath)
+      {
+        throw std::invalid_argument("Unable to determine entityPath.");
+      }
```

Comment:
This seems like a bit of a surprising behavior. Why would we only set the `m_entityPath` member field, when it is empty?

Who sets this field initially, given when the ctor is called, it would be empty anyway? It seems to be written in a way to support calling the ctor on the same object multiple times and get different behaviors.

Reply:
If the SAS credential string has an entitypath entry, we use the entity path from the credential string, otherwise we use the one the customer provided.

If the customer provided an entity path which is different from the one in the connection string, we will fail.